File: acn/custom_script/purchase_order/purchase_order.js
frappe.ui.form.on('Purchase Order', {
    refresh: function (frm) {
        // Hide/Show submit button

        // Add custom buttons only when doc is submitted
        if (frm.doc.open_order && frm.doc.reopen === 1) {

            frm.add_custom_button(__('Re-Close Order'), function () {
                frappe.call({
                    method: "acn.custom_script.purchase_order.purchase_order.reopen_order",
                    args: { name: frm.doc.name, docstatus: 1 }, // keep as submitted
                    callback: function (r) {
                        if (r.message) {
                            frappe.msgprint(r.message);
                            frm.reload_doc();
                        }
                    }
                });
            });
        }
        if (frm.doc.open_order && frm.doc.docstatus == 1) {
            frm.add_custom_button(__('Reopen Order'), function () {
                frappe.call({
                    method: "acn.custom_script.purchase_order.purchase_order.reopen_order",
                    args: { name: frm.doc.name, docstatus: 0 }, // set back to draft
                    callback: function (r) {
                        if (r.message) {
                            frappe.msgprint(r.message);
                            frm.reload_doc();
                        }
                    }
                });
            });
        }

    
}
});

File: acn/custom_script/purchase_order/purchase_order.py
import frappe
from frappe.model.document import Document



@frappe.whitelist()
def before_submit(self, method=None):
    if self.reopen:
        frappe.throw("This Purchase Order is reopened. You cannot submit it. Please click on Re-Close first.")
 

@frappe.whitelist()
def reopen_order(name,docstatus=0):
    """
    Reopen a submitted Purchase Order by setting flags
    """
    try:
        docstatus = int(docstatus)
        po = frappe.get_doc("Purchase Order", name)
        
      
        po.db_set("docstatus",docstatus)
        po.db_set("reopen", 0 if docstatus else 1)


        for d in  po.items:
            d.db_set("docstatus",docstatus)
        for d in  po.taxes:
            d.db_set("docstatus",docstatus)


        frappe.db.commit()

        return {"status": "success", "message": f"Purchase Order {name} reopened successfully."}
    
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Reopen Purchase Order Error")
        return {"status": "error", "message": str(e)}
